Source/Core/VideoCommon/PerformanceTracker.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <array>
+#include <chrono>
+#include <fstream>
+#include <mutex>
+#include <optional>
+
+#include "Common/CommonTypes.h"
+
+class PerformanceTracker
+{
+private:
+  // Must be powers of 2 for masking to work
+  static constexpr u64 MAX_DT_QUEUE_SIZE = 1UL << 12;
+  static constexpr u64 MAX_QUALITY_GRAPH_SIZE = 1UL << 8;
+
+  static inline std::size_t IncrementIndex(const std::size_t index)
+  {
+    return (index + 1) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+  static inline std::size_t DecrementIndex(const std::size_t index)
+  {
+    return (index - 1) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+  static inline std::size_t GetDifference(const std::size_t begin, const std::size_t end)
+  {
+    return (end - begin) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+private:
+  using Clock = std::chrono::steady_clock;
+  using TimePoint = Clock::time_point;
+  using RawDt = Clock::duration;
+
+  using MicroDt = std::chrono::duration<s64, std::micro>;
+  using MilliDt = std::chrono::duration<double, std::milli>;
+  using SecondsDt = std::chrono::duration<double>;
+
+public:
+  PerformanceTracker(const char* log_name = nullptr,
+                     const std::optional<s64> sample_window_us = {});
+  ~PerformanceTracker();
+
+  PerformanceTracker(const PerformanceTracker&) = delete;
+  PerformanceTracker& operator=(const PerformanceTracker&) = delete;
+  PerformanceTracker(PerformanceTracker&&) = delete;
+  PerformanceTracker& operator=(PerformanceTracker&&) = delete;
+
+public:  // Functions for recording and accessing information
+  void Reset();
+  void Count();
+
+  RawDt GetSampleWindow() const;
+  double GetSampleWindowMS() const;
+
+  double GetHzAvg() const;
+  double GetDtAvgMS() const;
+  double GetDtStdMS() const;
+
+  double GetLastRawDtS() const;
+
+  void ImPlotPlotLines(const char* label) const;
+
+private:  // Functions for managing dt queue
```

Comment:
Same here with the private/public.  Also it's helpful if you group the functions by access type (where possible).